lib/Dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp

```diff
@@ -80,21 +60,57 @@ CosimAccelerator::connect(std::string connectionString) {
   return std::make_unique<CosimAccelerator>(host, port);
 }
 
+struct esi::backends::cosim::CosimAccelerator::Impl {
+  capnp::EzRpcClient rpcClient;
+  kj::WaitScope &waitScope;
+  CosimDpiServer::Client cosim;
+  EsiLowLevel::Client lowLevel;
+
+  Impl(std::string hostname, uint16_t port)
+      : rpcClient(hostname, port), waitScope(rpcClient.getWaitScope()),
+        cosim(rpcClient.getMain<CosimDpiServer>()), lowLevel(nullptr) {
+    auto llReq = cosim.openLowLevelRequest();
+    auto llPromise = llReq.send();
+    lowLevel = llPromise.wait(waitScope).getLowLevel();
+  }
+};
+
 /// Construct and connect to a cosim server.
 // TODO: Implement this.
-CosimAccelerator::CosimAccelerator(std::string hostname, uint16_t port)
-    : info(nullptr) {
-  std::cout << hostname << ":" << port << std::endl;
-}
-CosimAccelerator::~CosimAccelerator() {
-  if (info)
-    delete info;
+CosimAccelerator::CosimAccelerator(std::string hostname, uint16_t port) {
+  impl = std::make_unique<Impl>(hostname, port);
 }
 
-const SysInfo &CosimAccelerator::sysInfo() {
-  if (info == nullptr)
-    info = new CosimSysInfo();
-  return *info;
+namespace {
+class CosimMMIO : public MMIO {
+public:
+  CosimMMIO(EsiLowLevel::Client &llClient, kj::WaitScope &waitScope)
+      : llClient(llClient), waitScope(waitScope) {}
+
+  uint64_t read(uint32_t addr) const override {
+    auto req = llClient.readMMIORequest();
+    req.setAddress(addr);
+    return req.send().wait(waitScope).getData();
+  }
+  void write(uint32_t addr, uint64_t data) override {
+    auto req = llClient.writeMMIORequest();
+    req.setAddress(addr);
+    req.setData(data);
+    req.send().wait(waitScope);
+  }
+
+private:
+  EsiLowLevel::Client &llClient;
+  kj::WaitScope &waitScope;
+};
+} // namespace
+
+Service *CosimAccelerator::createService(Service::Type svcType) {
+  if (svcType == typeid(MMIO))
+    return new CosimMMIO(impl->lowLevel, impl->waitScope);
+  else if (svcType == typeid(SysInfo))
+    return new MMIOSysInfo(getService<MMIO>());
+  return nullptr;
```

Comment:
To be frank, i don't think there is any merit in this kind of code returning a raw pointer. If you really want the raw pointer semantics, then there should be a `Service* getService` method that dispatches to a `std::unique_ptr<Service> createService` function in case the service was not found in the cache.

Reply:
That's what I do, but I wrap the pointer in another method.